### PR TITLE
feat: add config option: timeout for specific forwards

### PR DESCRIPTION
**Pull request by nritsche**

### CODE REVIEW

File: coco/request_forwarder.py [hunk 1]
```diff
@@ -58,15 +59,22 @@ async def trigger(self, method, request=None, hosts=None, params=None):
...
+            timeout=self.timeout,
         )
         if self.check:
             for check in self.check:
                 forward_result.success &= await check.run(forward_result)
 
         return forward_result
 
-    def forward_function(self, name, method, request, hosts=None, params=None):
+    def forward_function(
```
Comment (anjakefala) [inline coco/request_forwarder.py:75]: Why did method get moved?

**Comment by anjakefala** [comment]
From Rick about this PR:

> you can replace "forward" with "external endpoint". coco forwards requests to HTTP endpoints. coco has a timeout setting that it applies globally. But we need to be able to set a higher timeout just for get-frame0-time. that's an "external endpoint" on carillon/fpga-master. In the coco config it's called a forward. The coco config defines "endpoints" as in endpoints coco offers. And it will forward the requests internally according to the config.

Where this config option will get set: https://github.com/chime-experiment/ch_ansible/blob/master/roles/internal/coco_common/files/endpoints/gps-time.conf

File: coco/request_forwarder.py [hunk 2]
```diff
@@ -19,11 +19,12 @@
 class Forward:
     """Keep data about a forward to another endpoint."""
 
-    def __init__(self, name, group=None, request=None, check=None):
+    def __init__(self, name, group=None, request=None, check=None, timeout=None):
```
Comment (anjakefala) [inline coco/request_forwarder.py:22]: Can you add in-code documentation on the expected unit for timeout? I.e. is it in `seconds`?

File: coco/request_forwarder.py (hunk 1 above)
Comment (nritsche) [reply coco/request_forwarder.py:75]: So it can be ignored (`**_`) in the implementation for internal forwards.

File: coco/request_forwarder.py [hunk 3]
```diff
@@ -274,43 +282,48 @@ async def _request(self, session, method, host, endpoint, request, params):
...
             Request data to forward.
-        group : str or list(Host)
+        hosts : str or list(Host)
             Hosts to forward to or group name.
+        method : str
+            HTTP method.
         params : list of (key, value) pairs
             URL query parameters to forward to target endpoint.
+        timeout : int
+            Timeout in seconds. If none is supplied, the timeout from the top level of coco's config is used.
```
Comment (anjakefala) [inline coco/request_forwarder.py:302]: Oh, are the docs all here? =S

File: coco/check.py [hunk 4]
```diff
@@ -71,7 +71,7 @@ async def on_failure(self, hosts=None):
             logger.debug(f"Calling {self.on_failure_call} because {self._name} failed.")
             result.embed(
                 self.on_failure_call,
-                await self.forwarder.internal(self.on_failure_call, "", {}),
```
Comment (anjakefala) [inline coco/check.py:74]: Method got switched from `""` to `None`, do you need to do `if method is None: method = ""` anywhere? Or do the methods all know how to deal with `method` having a value of `None`?

**Review by anjakefala — COMMENTED** [review]
I had a request for documentation, and a few questions. Once docs are updated, and the question is answered, this looks good to me.

File: coco/check.py (hunk 4 above)
Comment (nritsche) [reply coco/check.py:74]: I don't think the `""` was used at all. `internal` ignores `method`

**Pull request merged**